### PR TITLE
Mm 68506 fe abac mask fe table editor cel and e2e

### PR DESCRIPTION
**Pull request by pvev**
#### Summary

This PR implements the frontend portion of attribute-value masking. When the backend hides specific attribute values from an admin who is not authorized to view them, the UI now:

- Displays a masked chip placeholder
- Switches affected editors into read-only mode
- Prevents operations that could unintentionally broaden access

This work builds on top of PR-A (backend) and covers four stories:

- **MM-68505** — Adds the `has_masked_values` type and introduces the `MaskedChip` component (the `--------` placeholder shown instead of hidden values).
- **MM-68506** — Updates the Table Editor and team settings modal:
  - masked rows render the chip
  - affected rows become read-only
  - Test/Delete actions are disabled when a rule contains hidden values
  - a banner explains why those actions are unavailable

- **MM-68507** — Updates the CEL editor to:
  - switch into read-only mode when masked rules are present
  - show a tooltip explaining the restriction

- **MM-68508** — Adds Playwright E2E coverage for:
  - masked chip rendering
  - read-only enforcement
  - save/delete flows
  - select, multiselect, and text-field scenarios

Server-side error IDs introduced in PR-A (`masked_condition_deleted`, `masked_rule_deleted`, `self_exclusion`, `invalid_value`) are now mapped to localized UI messages during save flows.

The policies list page was also updated to detect masked policies in the delete-confirmation modal using the actual server-emitted token, replacing an earlier stale check.

---

### Prerequisites

- Enterprise Advanced license
- Feature flags enabled:
  - `AttributeBasedAccessControl=true`
  - `AttributeValueMasking=true`
- At least one CPA field configured as `shared_only`

### Manual Verification

- As an admin who does **not** hold all values referenced by a policy:
  - open the editor
  - masked rows should display the chip
  - rows should be read-only
  - Test/Delete actions should be disabled with explanatory tooltips

- As an admin who **does** hold all values:
  - open the same policy
  - full read/write access should be available
  - no masked chips should appear

- Edit a non-masked part of a policy (e.g. channel assignments or name) and save:
  - save should succeed
  - hidden values should remain preserved in storage

- Attempt to delete a rule containing hidden values:
  - save should fail
  - localized “You cannot remove a rule…” message should appear

- Attempt to delete a condition containing hidden values:
  - save should fail
  - localized validation message should appear

- From the policies list page:
  - click Delete on a policy with masked values
  - confirmation modal should display the warning banner
  - deleting a non-masked policy should not show the banner

- CEL editor:
  - open a policy containing masked rules in CEL view
  - editor should be read-only with an explanatory tooltip

- Switch between Table and CEL editors while masked rules are present:
  - state should remain preserved
  - no flicker or hook errors should occur

## Security Considerations

This PR only implements the UI layer for a policy that is fully enforced server-side. A few important details worth calling out:

- The frontend is **not** the source of authority. Read-only editors, disabled buttons, and banners are strictly UX safeguards around backend enforcement introduced in PR-A. Even if someone bypasses the UI, the server still returns `403` responses for:
  - masked-condition removal
  - masked-rule removal
  - deletion attempts involving masked values

- The masked sentinel token (`--------`) is rendered only as a visual placeholder. The frontend never interprets or re-submits it as a real attribute value. Server-side merge logic strips it from incoming payloads.

- The policies list page detects masked policies using a substring check against raw CEL expressions. This is intentional because the list view only has access to raw expressions, not the parsed visual AST. In the unlikely case that a policy literally contains `"--------"` as a value, the warning banner may appear incorrectly, but the impact is purely cosmetic — no authorization decisions depend on this check.

- No new endpoints or permissions were introduced. All masking logic and authorization decisions remain server-side using existing handlers. The client only consumes `has_masked_values` and renders the appropriate UI state.

- System admins are **not exempt** from masking. If a sysadmin happens to hold all values referenced by a policy, masking will not appear during testing, which is expected behavior.
#### Ticket Link
https://mattermost.atlassian.net/browse/MM-68505
https://mattermost.atlassian.net/browse/MM-68506
https://mattermost.atlassian.net/browse/MM-68507

#### Screenshots

https://github.com/user-attachments/assets/b1889693-df86-4f75-a89c-e7536297bc1a

#### Release Note
```release-note
Added attribute value masking for access control policies: admins who don't hold specific attribute values now see a masked placeholder (--------) instead of the raw value, with affected rule editors switched to read-only mode and save/delete actions disabled. Includes masking support in the table editor, CEL editor, and team settings modal.
```

## Change Impact: 🔴 High

**Regression Risk:** This change touches both frontend UI editors and critical server-side access-control, masking, and expression-merge logic that affect policy enforcement and persistence; despite extensive tests, subtle edge cases in CEL round-tripping and masked-value merges could cause regressions across save/delete flows and multi-user visibility.  
**QA Recommendation:** Extensive manual QA recommended — focus on CEL ↔ Table round-trip, save/delete behavior with partially/fully masked rules, multi-user differential-visibility scenarios, and recovery paths for failed merges; automated tests provide good baseline coverage but targeted exploratory testing is necessary.

*Generated by CodeRabbitAI*

[Review Change Stack](https://app.coderabbit.ai/change-stack/mattermost/mattermost/pull/36517?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by pvev** [comment]
@coderabbitai review

File: webapp/channels/src/components/admin_console/access_control/editors/table_editor/masked_chip.tsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react';
+import {useIntl} from 'react-intl';
+
+import WithTooltip from 'components/with_tooltip';
```
[the remark this answers is not in the document]
Comment (pvev) [reply webapp/channels/src/components/admin_console/access_control/editors/table_editor/masked_chip.tsx]: sure, done!

File: e2e-tests/playwright/specs/functional/system_console/abac/masking/masking_db_setup.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+export function setFieldAsSharedOnly(fieldId: string): void {
+    const dbUrl = resolveDbUrl();
+    const sql = [
+        `UPDATE propertyfields`,
+        `SET attrs = jsonb_set(COALESCE(attrs, '{}'::jsonb), '{access_mode}', to_json('shared_only'::text)::jsonb),`,
+        `updateat = EXTRACT(EPOCH FROM NOW())::bigint * 1000`,
+        `WHERE id = '${fieldId}';`,
+    ].join(' ');
+
+    execSync(`psql "${dbUrl}" -c "${sql}"`, {stdio: 'pipe'});
```
[the remark this answers is not in the document]
Comment (pvev) [reply e2e-tests/playwright/specs/functional/system_console/abac/masking/masking_db_setup.ts]: nice suggestion, done.

File: server/channels/app/properties/access_control_field_test.go [hunk 3]
```diff
@@ -1440,7 +1440,3 @@ func TestLinkedPropertyField_SecurityInheritance(t *testing.T) {
 		assert.False(t, model.IsPropertyFieldProtected(linked))
 	})
 }
-
```
Comment (pvev) [inline server/channels/app/properties/access_control_field_test.go:1509]: not relevant comment deleted. Related to masking feature (went in one of the initial PRs)

File: webapp/channels/src/components/admin_console/access_control/editors/cel_editor/editor.tsx [hunk 4]
```diff
@@ -338,6 +340,19 @@ function CELEditor({
         <div className={`cel-editor ${className}`}>
             <MonacoLanguageProvider schemas={schemas}/>
 
+            {hasMaskedRows && (
```
Comment (pvev) [inline webapp/channels/src/components/admin_console/access_control/editors/cel_editor/editor.tsx:343]: visible in advanced editor, will be completely disabled in presence of masked vales.

File: webapp/channels/src/components/admin_console/access_control/editors/table_editor/masked_chip.tsx [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    return (
+        <WithTooltip title={tooltipText}>
+            <div
+                className='select__multi-value select__multi-value--masked'
+                role='img'
+                aria-label={ariaLabel}
+            >
+                <div className='select__multi-value__label'>
+                    {'••••••••'}
```
Comment (pvev) [inline webapp/channels/src/components/admin_console/access_control/editors/table_editor/masked_chip.tsx:33]: symbol to display, different from the one used in the server as constant. The string is more reliable in go code, the symbol works for UI display

File: webapp/channels/src/components/admin_console/access_control/policy_details/policy_details.tsx [hunk 6]
```diff
@@ -197,6 +198,14 @@ function PolicyDetails({
                 if (result.error) {
                     if (result.error.server_error_id === 'app.pap.save_policy.name_exists.app_error') {
                         setServerError(formatMessage({id: 'admin.access_control.edit_policy.name_exists', defaultMessage: 'A policy with this name already exists. Please choose a different name.'}));
+                    } else if (result.error.server_error_id === 'app.pap.save_policy.invalid_value') {
```
Comment (pvev) [inline webapp/channels/src/components/admin_console/access_control/policy_details/policy_details.tsx:210]: I do not like these many if-else, will open a subsequent PR with a focused clean up for the policy details component. There is lot of room for improvement, not just this. But will work on a focused PR for it after this.

File: webapp/channels/src/components/admin_console/access_control/policies.tsx [hunk 7]
```diff
@@ -157,10 +168,38 @@ export default function PolicyList(props: Props): JSX.Element {
         );
     };
 
-    const handleDelete = async (policyId: string) => {
-        await props.actions.deletePolicy(policyId);
+    const initiateDelete = useCallback((policy: AccessControlPolicy) => {
```
Comment (pvev) [inline webapp/channels/src/components/admin_console/access_control/policies.tsx:175]: this is UI UX experience, the server itself also blocks the deletion of policies with masked values. So, this acts as a good UX but also as the initial layer of defense

**Comment by lindalumitchell** [comment]
/spinwick update --env MM_FEATUREFLAGS_AttributeValueMasking=true

**Comment by lindalumitchell** [comment]
@pvev without db access on the spinwick, is there another way to accomplish `At least one CPA field configured as shared_only` for testing?

**Comment by pvev** [comment]
> @pvev without db access on the spinwick, is there another way to accomplish `At least one CPA field configured as shared_only` for testing?

hey @lindalumitchell , so, I created a QA plugin that is already uploaded to the spinwick server which will allow you to made some modifications to the CPA via a script (sent to you in DM abac-qa.sh). I already tried and is working fine (check Program attribute. User-1 is also an admin, if you check the policy with him, will see how the masked values are shown)

Screenshot 2026-05-16 at 13 14 49

**Comment by lindalumitchell** [comment]
@pvev thank you! Super helpful. 🙏 

EDIT: I can't repro the below anymore; possible it's an error in my testing, and I just hadn't refreshed after making a change. If you can't repro this, let's disregard for now.

Question I just came across: Is it expected that the user who sees a masked row also sees a different operator displayed on that row? Screenshot below shows `sysadmin` on the left, who has the attributes and can see the whole row (sees the operator `has any of`), and shows `user-1` on the right, who does not have the attributes and sees the values masked (sees the operator `has all of`).

Screenshot 2026-05-17 at 5 22 38 PM

**Comment by lindalumitchell** [comment]
This one is a real question though :)

A user who sees masking in a policy attempts to delete the policy from the policy list page. They see a confirmation modal that includes a warning about the masked value(s), as I believe is expected. They confirm the delete, but the policy is not removed from the list, even after refresh.

Then a user who does not see masking in that same policy attempts to delete that policy from the policy list page. They are able to delete without issue (though it does take a refresh to remove it from the UI):

https://github.com/user-attachments/assets/f3c2b0a4-913a-4c0d-a97c-1f6e2f893555

**Comment by pvev** [comment]
@amyblais can we get this one cherry-picked for 11.8 too? This one is the FE part of https://github.com/mattermost/mattermost/pull/36513 and both are behind a feature-flag

**Comment by pvev** [comment]
> This one is a real question though :)
> 
> A user who sees masking in a policy attempts to delete the policy from the policy list page. They see a confirmation modal that includes a warning about the masked value(s), as I believe is expected. They confirm the delete, but the policy is not removed from the list, even after refresh.
> 
> Then a user who does not see masking in that same policy attempts to delete that policy from the policy list page. They are able to delete without issue (though it does take a refresh to remove it from the UI):
> 
>  Screen.Recording.2026-05-17.at.5.38.14.PM.mp4

Thanks @lindalumitchell , good catch! , I fixed this on my last commit.

**Comment by amyblais** [comment]
> can we get this one cherry-picked for 11.8 too? This one is the FE part of #36513 and both are behind a feature-flag

@pvev Yes 👍 I hadn't added the "cherry-picked approved" label earlier because this PR isn't submitted against the master branch.

**Comment by pvev** [comment]
> > can we get this one cherry-picked for 11.8 too? This one is the FE part of #36513 and both are behind a feature-flag
> 
> @pvev Yes 👍 I hadn't added the "cherry-picked approved" label earlier because this PR isn't submitted against the master branch.

thank you @amyblais 🚀

File: server/channels/api4/access_control.go
Comment (davidkrauser) [inline server/channels/api4/access_control.go:145]: Could this be called inside the `Update/Search/Get/etc.AccessControlPolicy()` methods themselves? That way we don't forget to call them? The fewer places we need to remember to call these, the better.

File: server/channels/app/access_control.go
Comment (davidkrauser) [inline server/channels/app/access_control.go:380]: Why was this removed? What did it do?

Comment (davidkrauser) [inline server/channels/app/access_control.go:292]: Is `ExpressionToVisualAST()` an expensive operation? We're using it a lot, and wondering if it make sense to cache the results (e.g. a map of expressions -> visual ASTs)? (0/5)

Comment (davidkrauser) [inline server/channels/app/access_control.go:283]: This is a big complex function. It might help with readability/testability if we break it up into smaller pieces.

**Comment by pvev** [comment]
> @pvev thank you! Super helpful. 🙏
> 
> EDIT: I can't repro the below anymore; possible it's an error in my testing, and I just hadn't refreshed after making a change. If you can't repro this, let's disregard for now.
> 
> Question I just came across: Is it expected that the user who sees a masked row also sees a different operator displayed on that row? Screenshot below shows `sysadmin` on the left, who has the attributes and can see the whole row (sees the operator `has any of`), and shows `user-1` on the right, who does not have the attributes and sees the values masked (sees the operator `has all of`).
> 
> Screenshot 2026-05-17 at 5 22 38 PM

thank you @lindalumitchell . Fixed in my last commit. 


https://github.com/user-attachments/assets/43a168d7-2570-48a1-a67f-8ae0f2d841b2

File: server/channels/app/access_control_masking.go
Comment (davidkrauser) [inline server/channels/app/access_control_masking.go:867]: On line 839 above, we're setting the expression to "true" if there's an app error. What happens if a caller who holds all the values re-submits the policy with no changes? Would that replace the expression with "true" in the backend?

**Review by davidkrauser — COMMENTED** [review]
I think you were fixing up some merge conflicts or pulling some things out, so my review comments got a little messy. I had two more comments for `access_control.go` that I couldn't place inline:

1. We don't compare the Rules.Actions is the same before overwriting with something provided by the caller, so someone could change the Actions on a Rule that they should not be able to modify. Consider this scenario:

```
  Rules: [
    { Actions: ["membership"], Expression: 'user.attributes.clearance == "TopSecret"' }
  ]
```

  The caller cannot see "TopSecret". What they GET'd back looks like:

```
  Rules: [
    { Actions: ["membership"], Expression: 'user.attributes.clearance == "--------"' }
  ]
```

  They submit a PUT/POST with the action swapped, expression unchanged:
  
```
  Rules: [
    { Actions: ["upload_file_attachment"], Expression: 'user.attributes.clearance == "--------"' }
  ]
```

I think this now means that anyone can join the channel (but uploading files will be restricted).

2. We aren't masking values coming back from `/api/v4/channels/{channel_id}/access_control/attributes`, so we're leaking data there. Were we supposed to mask those in this effort? I'm not sure if there are other instances that need to be masked, too.

File: server/channels/app/access_control.go
Comment (davidkrauser) [reply server/channels/app/access_control.go:283]: This is a low-impact nitpick - dropping

Comment (davidkrauser) [reply server/channels/app/access_control.go:380]: This is a low-impact nitpick. Dropping

**Review by davidkrauser — CHANGES_REQUESTED** [review]

**Review by lindalumitchell — APPROVED** [review]
Verified that the couple questions I had were resolved; thanks @pvev ! Just to note here that the expected behavior (which is true now) is that a user who sees masked value(s) on a policy cannot delete that policy.

E2es are in place and thorough! LGTM 👍 

AI summary for my reference:

**QA Approved** ✅

Manually tested attribute-value masking across the ABAC policy UI per the [QA test plan](https://mattermost.atlassian.net/wiki/x/CoDxDwE). Coverage included:

- **Feature flag gating** — masking applies when `AttributeValueMasking = true`; all controls remain unrestricted when OFF
- **Masked chip UI** — `••••••••` chip renders correctly with `role="img"`, correct `aria-label`, tooltip, no remove button, and always positioned after visible chips
- **Read-only controls on masked rows** — attribute, operator, and value selectors all locked; row remove button disabled (`opacity: 0.4`, `cursor: not-allowed`)
- **Warning banner** — appears when masked values are present, absent for clean policies, and persists across Simple ↔ Advanced mode toggles without flickering
- **CEL editor** — read-only with restricted-values notice when masked; editable when caller holds all values; sentinel token (`"--------"`) shown in place of non-held values
- **Test access rule button** — disabled (with correct tooltip) when masked values present; enabled for full-visibility callers
- **Save / merge-on-save** — hidden values preserved in DB after save by a masked caller; confirmed via direct DB query for both partial and fully-masked conditions, including multi-condition policies
- **Server-side write guards** — HTTP 400 for non-held or sentinel literal values via direct API; HTTP 403 for dropping masked conditions or rows
- **Self-exclusion** — save blocked with correct error when admin removes their only satisfying value
- **Delete flows** — delete button disabled in policy editor and team settings modal when masked values present; server returns HTTP 403 on direct DELETE; standard confirmation modal shown (no spurious warning) for clean policies; delete button stays disabled after removing all channel assignments
- **New policy creation** — no masked state on a fresh policy form
- **Regression** — non-masked CRUD workflows unaffected; Save Changes panel button CSS classes correct

File: server/channels/api4/access_control.go
Comment (pvev) [reply server/channels/api4/access_control.go:145]: Valid pint from architecturally perspective, but not practical here. But the thing is shouldRedactExpressions(c) takes *Context api layer and the app methods don't have it. Pushing masking down would mean either changing app method signatures to accept like "should redact" flag, or duplicating feature flag / environment checks in the app layer. So, 3/5 keeping it as is.

File: server/channels/app/access_control.go
Comment (pvev) [reply server/channels/app/access_control.go:292]: well, maybe? at least for the scope of this PR this is a save path, called once per PUT. Not a hot loop, not a search/list path. In any case, is worth noting MaskPolicyExpressions (GET/search) already minimizes cost within each call 👍

File: server/channels/app/access_control_masking.go
Comment (pvev) [reply server/channels/app/access_control_masking.go:867]: No risk, MaskPolicyExpressions only mutates the in-memory struct for the response and never touches the DB. The fail-closed "true" only fires if the  stored expression can't be parsed. On the save path, mergeExpressionWithMaskedValues calls expressionHasMaskedValuesForCaller first, which also parses the stored expression and only then if it's "unparseable" there too, the save returns an error before anything is written. So a re-submitted "true" would be rejected, not persisted. In any case added a unit test to cover this TestMergeStoredPolicyExpressions_FailClosedTrueRejectedOnResubmit . Added to test covering the parse-failure path.

**Comment by pvev** [comment]
> I think you were fixing up some merge conflicts or pulling some things out, so my review comments got a little messy. I had two more comments for `access_control.go` that I couldn't place inline:
> 
> 1. We don't compare the Rules.Actions is the same before overwriting with something provided by the caller, so someone could change the Actions on a Rule that they should not be able to modify. Consider this scenario:
> 
> ```
>   Rules: [
>     { Actions: ["membership"], Expression: 'user.attributes.clearance == "TopSecret"' }
>   ]
> ```
> 
> The caller cannot see "TopSecret". What they GET'd back looks like:
> 
> ```
>   Rules: [
>     { Actions: ["membership"], Expression: 'user.attributes.clearance == "--------"' }
>   ]
> ```
> 
> They submit a PUT/POST with the action swapped, expression unchanged:
> 
> ```
>   Rules: [
>     { Actions: ["upload_file_attachment"], Expression: 'user.attributes.clearance == "--------"' }
>   ]
> ```
> 
> I think this now means that anyone can join the channel (but uploading files will be restricted).
> 
> 2. We aren't masking values coming back from `/api/v4/channels/{channel_id}/access_control/attributes`, so we're leaking data there. Were we supposed to mask those in this effort? I'm not sure if there are other instances that need to be masked, too.

Thanks for the feedback @davidkrauser , so answering to these two:
1) This was a real gap. Good catch, thank you! so, Fix was straightforward: after mergeing the expression, if the stored rule had masked values for this caller, lock 'Actions' back to the stored rule's value.

2) Good call. Solved by added a filter in GetAccessControlPolicyAttributes that looks up each fields access mode after getting the raw attributes from the enforcement cache so source_only and shared_only fields are deleted before the response goes out. Also, the endpoint fails closed: unresolvable field == omit. Public fields pass unchanged. The channel info banner and invite modal will just not show those attribute tags at all for masked fields

**Review by abhijit-singh — APPROVED** [review]
Great work @pvev! Just some minor visual requests and copy changes:
1. For the second warning message shown in the advanced mode, it is conflicting a bit with the section notice above it. Can we update the copy and style it closer to this (Figma):
**Current**

Image


**Proposed**
Image


2. For the disabled state in the simple mode, the chips have different padding on the left and right of the text. Can we increase the right one to make it equal to the left?

Image

3. Instead of showing a section notice in the delete section also, can we show a tooltip on the disabled delete button similar to how we have it for the `Test access rule` button?

Image

4. Can we change the body copy for the section notice:
    - Current: "Some rules include attribute values you cannot see. Editing or deleting these rules may change who has access in ways you cannot fully anticipate."
    - Proposed: "Some rules include attribute values you cannot see. Editing or deleting these rules may change access in unpredictable ways."

**Review by davidkrauser — APPROVED** [review]
LGTM - thank you for making changes 👍

**Comment by pvev** [comment]
@saturninoabril any idea why the e2e job is failing even when the 100% of the tests passed?. Thanks 🙏

**Comment by pvev** [comment]
/test-analysis-override we already discussed the relevance of the not added tests in our previous conversation, and for the e2e test it will immediately follow this PR once figure out what exactly is causing the issue

**Comment by pvev** [comment]
will open a new PR with the e2e tests changes and handle the configuration issues there

**Comment by saturninoabril** [comment]
> any idea why the e2e job is failing even when the 100% of the tests passed?. Thanks 🙏

It failed on the following. I'd suggests to break the tests into multiple spec files (then each spec file to be within 2 mins overall tests only so that when need to retest, it's narrowed down to some sets of test only and faster retest duration).
Screenshot 2026-05-20 at 10 49 13 AM

```
Running 23 tests using 1 worker

  ✘   1 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:327:9 › Attribute-Value Masking › MM-68508-1: Full masking round-trip in Simple editor (36.5s)
  ✘   2 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:327:9 › Attribute-Value Masking › MM-68508-1: Full masking round-trip in Simple editor (retry #1) (35.5s)
  ✘   3 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:410:9 › Attribute-Value Masking › MM-68508-2: Caller with masked values can save; hidden values are preserved by merge (37.0s)
  ✘   4 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:410:9 › Attribute-Value Masking › MM-68508-2: Caller with masked values can save; hidden values are preserved by merge (retry #1) (33.2s)
  ✘   5 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:493:9 › Attribute-Value Masking › MM-68508-3: Row-remove button is disabled on masked rows (36.5s)
  ✘   6 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:493:9 › Attribute-Value Masking › MM-68508-3: Row-remove button is disabled on masked rows (retry #1) (36.4s)
  ✓   7 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:559:9 › Attribute-Value Masking › MM-68508-4: Self-inclusion failure blocks save (caller has full visibility) (34.1s)
  ✓   8 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:640:9 › Attribute-Value Masking › MM-68508-5: Non-held value rejected via direct API (11.1s)
  ✘   9 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:717:9 › Attribute-Value Masking › MM-68508-6: CEL editor is read-only when policy has masked values (36.0s)
  ✘  10 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:717:9 › Attribute-Value Masking › MM-68508-6: CEL editor is read-only when policy has masked values (retry #1) (36.7s)
  ✓  11 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:813:9 › Attribute-Value Masking › MM-68508-7: Caller holding all policy values sees them all unmasked (27.8s)
  ✓  12 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:895:9 › Attribute-Value Masking › MM-68508-8: New policy creation has no masking (18.3s)
  ✘  13 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:951:9 › Attribute-Value Masking › MM-68508-9: Masked row is fully read-only; merge-on-save preserves hidden values (37.4s)
  ✘  14 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:951:9 › Attribute-Value Masking › MM-68508-9: Masked row is fully read-only; merge-on-save preserves hidden values (retry #1) (33.4s)
  ✘  15 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:1043:9 › Attribute-Value Masking › MM-68508-10: Text field masking with "in" operator (36.5s)
  ✘  16 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:1043:9 › Attribute-Value Masking › MM-68508-10: Text field masking with "in" operator (retry #1) (36.0s)
  ✘  17 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:1105:9 › Attribute-Value Masking › MM-68508-11: Text field masking with single-value operator (value not held) (31.2s)
  ✘  18 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:1105:9 › Attribute-Value Masking › MM-68508-11: Text field masking with single-value operator (value not held) (retry #1) (30.7s)
  ✘  19 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:1169:9 › Attribute-Value Masking › MM-68508-12: GET /policies/{id} does not leak raw CEL when values are masked (25.4s)
  ✘  20 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:1169:9 › Attribute-Value Masking › MM-68508-12: GET /policies/{id} does not leak raw CEL when values are masked (retry #1) (25.3s)
  ✘  21 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:1240:9 › Attribute-Value Masking › MM-68508-13: POST /policies/search does not leak raw CEL when values are masked (21.6s)
  ✘  22 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:1240:9 › Attribute-Value Masking › MM-68508-13: POST /policies/search does not leak raw CEL when values are masked (retry #1) (21.2s)
  ✘  23 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:1303:9 › Attribute-Value Masking › MM-68508-14: Warning banner visible in editor when policy has masked values (36.9s)
  ✘  24 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:1303:9 › Attribute-Value Masking › MM-68508-14: Warning banner visible in editor when policy has masked values (retry #1) (38.9s)
  ✘  25 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:1371:9 › Attribute-Value Masking › MM-68508-15: Delete button is disabled on masked policies; clean policies open the standard confirmation modal (41.2s)
  ✘  26 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:1371:9 › Attribute-Value Masking › MM-68508-15: Delete button is disabled on masked policies; clean policies open the standard confirmation modal (retry #1) (41.2s)
  ✘  27 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:1456:9 › Attribute-Value Masking › MM-68508-16: Delete Policy is blocked (UI and server) when caller has masked values (34.0s)
  ✘  28 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:1456:9 › Attribute-Value Masking › MM-68508-16: Delete Policy is blocked (UI and server) when caller has masked values (retry #1) (33.7s)
  ✘  29 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:1541:9 › Attribute-Value Masking › MM-68508-17: Multi-condition save preserves all hidden values; deleting masked row is blocked (35.4s)
  ✘  30 [chrome] › specs/functional/system_console/abac/masking/attribute_value_masking.spec.ts:1541:9 › Attribute-Value Masking › MM-68508-17: Multi-condition save preserves all hidden values; deleting masked row is blocked (retry #1) (38.1s)
Error: The operation was canceled.
```
